crates/core/src/network/connection_manager.rs

```diff
@@ -632,41 +642,43 @@ where
     drop(connection);
 
     let drain_start = std::time::Instant::now();
-    trace!("Draining connection");
+    debug!("Draining connection");
     let mut drain_counter = 0;
-    // Draining of incoming queue
-    while let Some(Ok(msg)) = incoming.next().await {
-        // ignore malformed messages
-        let Some(header) = msg.header else {
-            continue;
-        };
-        if let Some(body) = msg.body {
-            // we ignore non-deserializable messages (serde errors, or control signals in drain)
-            if let Ok(msg) = body.try_as_binary_body(protocol_version) {
-                drain_counter += 1;
-                let parent_context = header.span_context.as_ref().map(|span_ctx| {
-                    global::get_text_map_propagator(|propagator| propagator.extract(span_ctx))
-                });
-
-                if let Err(e) = router
-                    .call(
-                        Incoming::from_parts(
-                            msg,
-                            // This is a dying connection, don't pass it down.
-                            WeakConnection::new_closed(peer_node_id),
-                            header.msg_id,
-                            header.in_response_to,
-                            PeerMetadataVersion::from(header),
+    if needs_drain {
+        // Draining of incoming queue
+        while let Some(Some(Ok(msg))) = incoming.next().now_or_never() {
```

Comment:
The `ConnectionAwareRpcRouter` on the sending end of this stream is keeping it open because it can hold the sender for this stream in a future it created and which is being awaited on the other end of the connection. Did I understand it correctly? Must have been quite hard to track this down. Respect.

Reply:
Correct